pkg/pool-weighted/contracts/LBPool.sol

```diff
@@ -0,0 +1,294 @@
+
+        // solhint-disable-next-line not-rely-on-time
+        uint32 currentTime = block.timestamp.toUint32();
+        _startGradualWeightChange(currentTime, currentTime, params.normalizedWeights, params.normalizedWeights);
+        _setSwapEnabled(swapEnabledOnStart);
+    }
+
+    /**
+     * @notice Return start time, end time, and endWeights as an array.
+     * @dev Current weights should be retrieved via `getNormalizedWeights()`.
+     * @return startTime The starting timestamp of any ongoing weight change
+     * @return endTime The ending timestamp of any ongoing weight change
+     * @return endWeights The "destination" weights, sorted in token registration order
+     */
+    function getGradualWeightUpdateParams()
+        external
+        view
+        returns (uint256 startTime, uint256 endTime, uint256[] memory endWeights)
+    {
+        PoolState memory poolState = _poolState;
+
+        startTime = poolState.startTime;
+        endTime = poolState.endTime;
+
+        endWeights = new uint256[](_NUM_TOKENS);
+        endWeights[0] = poolState.endWeight0;
+        endWeights[1] = FixedPoint.ONE - poolState.endWeight0;
+    }
+
+    /**
+     * @notice Indicate whether swaps are enabled or not for the given pool.
+     * @return swapEnabled True if trading is enabled
+     */
+    function getSwapEnabled() external view returns (bool) {
+        return _getPoolSwapEnabled();
+    }
+
+    /*******************************************************************************
+                                Permissioned Functions
+    *******************************************************************************/
+
+    /**
+     * @notice Enable/disable trading.
+     * @dev This is a permissioned function that can only be called by the owner.
+     * @param swapEnabled True if trading should be enabled
+     */
+    function setSwapEnabled(bool swapEnabled) external onlyOwner {
+        _setSwapEnabled(swapEnabled);
+    }
+
+    /**
+     * @notice Start a gradual weight change. Weights will change smoothly from current values to `endWeights`.
+     * @dev This is a permissioned function that can only be called by the owner.
+     * If the `startTime` is in the past, the weight change will begin immediately.
+     *
+     * @param startTime The timestamp when the weight change will start
+     * @param endTime  The timestamp when the weights will reach their final values
+     * @param endWeights The final values of the weights
+     */
+    function updateWeightsGradually(
+        uint256 startTime,
+        uint256 endTime,
+        uint256[] memory endWeights
+    ) external onlyOwner {
+        InputHelpers.ensureInputLengthMatch(_NUM_TOKENS, endWeights.length);
+
+        if (endWeights[0] < _MIN_WEIGHT || endWeights[1] < _MIN_WEIGHT) {
+            revert MinWeight();
+        }
+        if (endWeights[0] + endWeights[1] != FixedPoint.ONE) {
+            revert NormalizedWeightInvariant();
```

Comment:
This is a more global point, but it'd be great if we had errors that are more descriptive in their naming. This is very clear in context, but much less clear as an integrator.

Reply:
Legacy v2 again :) (v1 had denormalized weights.) We could review all the error messages / ask partners/auditors if any are confusing.

Reply:
marking resolved and adding as a comment to https://github.com/balancer/balancer-v3-monorepo/pull/971